src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingSubmissionService.java

```diff
@@ -165,26 +171,54 @@ public ProgrammingSubmission notifyPush(Long participationId, Object requestBody
         programmingSubmission.setCommitHash(commit.getCommitHash());
-        log.info("create new programmingSubmission with commitHash: " + commit.getCommitHash() + " for participation " + participationId);
+        log.info("Create new programmingSubmission with commitHash: " + commit.getCommitHash() + " for participation " + participationId);
 
         programmingSubmission.setSubmitted(true);
         programmingSubmission.setSubmissionDate(ZonedDateTime.now());
         programmingSubmission.setType(SubmissionType.MANUAL);
-
         programmingExerciseParticipation.addSubmission(programmingSubmission);
 
+        // Students are not allowed to submit a programming exercise after the exam due date, if this happens we set the Submission to ILLEGAL
+        checkForIllegalExamSubmission(programmingExerciseParticipation, programmingSubmission);
+
         programmingSubmission = programmingSubmissionRepository.save(programmingSubmission);
         // NOTE: we don't need to save the participation here, this might lead to concurrency problems when doing the empty commit during resume exercise!
         return programmingSubmission;
     }
 
+    /**
+     * We check if a submission for an exam programming exercise is after the individual end date and a student is not allowed to submit anymore.
+     * If this is the case, the submission is set to {@link SubmissionType#ILLEGAL}.
+     *
+     * @param programmingExerciseParticipation current participation of the exam exercise
+     * @param programmingSubmission            new created submission of the repository commit
+     */
+    private void checkForIllegalExamSubmission(ProgrammingExerciseParticipation programmingExerciseParticipation, ProgrammingSubmission programmingSubmission) {
+        ProgrammingExercise programmingExercise = programmingExerciseParticipation.getProgrammingExercise();
+        boolean isExamExercise = programmingExercise.isExamExercise();
+        // Students are not allowed to submit a programming exercise after the exam due date, if this happens we set the Submission to ILLEGAL
+        if (isExamExercise && programmingExerciseParticipation instanceof ProgrammingExerciseStudentParticipation) {
+            var optionalStudent = ((ProgrammingExerciseStudentParticipation) programmingExerciseParticipation).getStudent();
+            Optional<User> optionalStudentWithGroups = optionalStudent.isPresent() ? userRepository.findOneWithGroupsAndAuthoritiesByLogin(optionalStudent.get().getLogin())
+                    : Optional.empty();
+            if (optionalStudentWithGroups.isPresent() && !examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, optionalStudentWithGroups.get())) {
+                final String message = "The student " + optionalStudentWithGroups.get().getLogin()
+                        + "just illegally submitted code after the allowed individual due date (including the grace period) in the participation "
```

Comment:
I guess here you could add a space 
```suggestion
                        + " just illegally submitted code after the allowed individual due date (including the grace period) in the participation "
```